### PR TITLE
[BEAM-27] Add timerId to TimerData

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This `timerId` is generated to be identical to historical behavior, and to be unique per time domain and timestamp.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @bjchambers

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java [hunk 1]
```diff
@@ -63,12 +63,11 @@
   void setTimer(StateNamespace namespace, String timerId, Instant target, TimeDomain timeDomain);
 
   /**
-   * Writes out a timer to be fired when the watermark reaches the given timestamp, automatically
-   * generating an id for it from the provided {@link TimerData}.
+   * Writes out a timer to be fired when the watermark reaches the timestamp contained in the
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: "Writes ..." seems a bit odd. Maybe just "Sets the timer described by {@code timerKey}"

Also, I think timerKey should be renamed to timerData?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java [hunk 2]
```diff
@@ -78,7 +77,7 @@
   void deleteTimer(StateNamespace namespace, String timerId);
 
   /**
-   * Deletes the given timer, automatically inferring its ID from the {@link TimerData}.
+   * Deletes the timer with the ID contained in the provided {@link TimerData}.
    */
   void deleteTimer(TimerData timerKey);
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: `timerKey` -> `timerData`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java [hunk 3]
```diff
@@ -164,16 +163,31 @@
...
   class TimerData implements Comparable<TimerData> {
+    private final String timerId;
     private final StateNamespace namespace;
     private final Instant timestamp;
     private final TimeDomain domain;
 
-    private TimerData(StateNamespace namespace, Instant timestamp, TimeDomain domain) {
+    private TimerData(
+        StateNamespace namespace, String timerId, Instant timestamp, TimeDomain domain) {
+      this.timerId = timerId;
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: Order here should match order of parameters and order of fields?

`@AutoValue`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java (hunk 1 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java (hunk 3 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java]: Done. Nice.

**Pull request closed without merging**